### PR TITLE
fix: make claude-code provider work on Windows

### PR DESCRIPTION
**Pull request by papisoV**
## Summary

The claude-code provider is unusable on Windows out of the box. Running the full map/dig pipeline on Windows 11 surfaced four distinct failures. Three are fixed here, in `src/llm/claude-code.ts`; the fourth (repairing unbalanced JSON actions) has moved to #105 so this PR stays transport-only.

1. **spawn ENOENT** — `spawn("claude")` fails on Windows because the CLI is a `.cmd` shim. Windows now launches `claude.cmd` through a shell. Since a `.cmd` shim cannot be spawned without `shell: true`, and Node refuses to escape the argv it concatenates into a shell command line (DEP0190), the Windows command line is assembled explicitly with cmd.exe quoting rather than relying on a shell to escape it.

2. **System/user role boundary** — the rendered system prompt now travels by file via `--append-system-prompt-file`, and only the user turn is piped to stdin. It is written to the absolute path inside the per-call temp directory but referenced from the command line by the constant relative name `system-prompt.txt`, with the child's `cwd` set to that directory: an OS-derived absolute path on that line is a `%NAME%` expansion surface in cmd.exe, and the relative name has none.

3. **CreateProcess command-line limit** — long system prompts passed as a command-line argument exceed the Windows limit: `claude.cmd` exits instantly with code 255 and no output, which upstream surfaces as a 15-minute timeout. Passing the prompt by file (2) removes the prompt from the command line entirely.

4. **Injection surface** — `input.model` is the one caller-supplied value that reaches that command line. It is validated against `^[A-Za-z0-9][A-Za-z0-9._:@/-]*$` before the temp directory is created, so an unsafe id can neither reach cmd.exe nor leave a directory behind on its way out.

5. **Temp cleanup** — the hand-rolled retry loop is replaced by `fs.rm`'s own retries (`maxRetries: 5`, `retryDelay: 500`), and a persistent failure is reported instead of thrown, so it cannot mask the completion error already propagating out. Only bounded, non-identifying fields are logged — the error `code` and `syscall`; the directory is environment-derived and the raw message repeats it.

## Test plan

- [x] **`tests/claude-code-provider.test.mjs` (new, 9 tests).** `buildClaudeCommand(args, platform = process.platform)` takes the platform explicitly, so the Windows execution path is covered deterministically on every platform — no Windows CI job required. Covered: command construction per platform; cmd.exe quoting including spaces, embedded quotes and trailing backslashes; unsafe model ids rejected (both directly and through the invocation builder, i.e. before the temp dir exists); the `--append-system-prompt-file` argument is the constant relative name with no separator, drive letter or `%`; the system prompt never appears on the command line and only the user turn reaches stdin; cleanup success is silent, and a persistent failure is reported rather than thrown with exactly `{ code, syscall }` logged, on both the logger and the stderr branches.
- [x] **No regressions.** Full suite on Windows 11 with this change: 546 tests, 519 pass, 27 fail. Against the same commit without it: 537 tests, 510 pass, 27 fail. The failure sets are byte-identical (Apple container, Docker, toolchain pinning, provider auth, Windows path handling); the +9 are the new file.
- [x] **Live: the relative prompt path works and lands in the system role.** The model recalls a nonce that exists only in the prompt file, and a constraint placed in that file at system level is honored 3/3. (A "tell me which role you received this in" probe is *not* usable here — repeated, it answers `USRROLE` about two thirds of the time regardless of how the file is passed.)
- [x] **End-to-end map run** over `fixtures/harness-evolution/case-a` on Windows 11: completes autonomously, 10+ scored scopes, clean `done`.
- [ ] **Maintain behavior on macOS/Linux.** All Windows-specific behavior is gated on `process.platform === "win32"`; the POSIX path still spawns `claude` with `shell: false`. Not exercised on a POSIX host by me.

### CODE REVIEW

File: src/llm/claude-code.ts [hunk 1]
```diff
@@ -36,18 +36,22 @@ export class ClaudeCodeClient implements LlmClient {
...
       DISABLED_CLAUDE_TOOLS,
       "--permission-mode",
       "default",
     ];
+    // Windows CreateProcess has a 32k-char command-line limit; long system
+    // prompts passed via --append-system-prompt make claude.cmd exit(255)
+    // instantly with no output (upstream then sees a timeout). Fold the
+    // rendered system prompt into stdin instead — same semantics for a pure
+    // text-completion backend.
+    const stdinPayload = `${system}\n\n---\n\n${input.user}`;
```
Comment (adshao) [inline src/llm/claude-code.ts]: [P1] This is not equivalent to `--append-system-prompt`: both `system` and `input.user` now arrive as one user message on stdin. Flounder's response contract, tool protocol, and audit safety/evidence rules therefore lose their system-level priority on every platform, not only Windows. Please keep the role boundary and avoid the command-line limit by writing the rendered prompt to the temporary directory and passing `--append-system-prompt-file` (or another transport that preserves the system role).

File: src/llm/claude-code.ts [hunk 2]
```diff
@@ -131,7 +148,8 @@ function parseClaudeOutput(stdout: string): { text: string; meta: Record<string,
 
 function spawnClaude(args: string[], input: string, options: { cwd: string; maxBuffer: number; timeout: number }): Promise<string> {
   return new Promise((resolve, reject) => {
-    const child = spawn("claude", args, { cwd: options.cwd, stdio: ["pipe", "pipe", "pipe"] });
+    const isWindows = process.platform === "win32";
+    const child = spawn(isWindows ? "claude.cmd" : "claude", args, { cwd: options.cwd, stdio: ["pipe", "pipe", "pipe"], shell: isWindows });
```
Comment (adshao) [inline src/llm/claude-code.ts]: [P1] `shell: true` sends this through `cmd.exe` while `args` contains the configurable `input.model`. Non-pi provider model IDs are accepted as arbitrary strings by `readProviderInput`, so shell metacharacters can reach the command interpreter. Node explicitly warns against passing unsanitized input to `spawn` with a shell. Please launch without a shell by resolving the underlying executable/shim safely, or introduce strict validation plus correct Windows argument handling before invoking `cmd.exe`.

File: src/llm/claude-code.ts [hunk 3]
```diff
@@ -75,7 +79,20 @@ export class ClaudeCodeClient implements LlmClient {
...
       throw new Error(`claude-code completion failed: ${message}`);
     } finally {
-      await rm(tmp, { recursive: true, force: true });
+      // Windows: the spawned claude.cmd may still hold the temp cwd briefly
+      // after close (EBUSY) — retry instead of failing the whole call.
+      for (let attempt = 0; attempt < 5; attempt++) {
+        try {
+          await rm(tmp, { recursive: true, force: true });
+          break;
+        } catch (error) {
```
Comment (adshao) [inline src/llm/claude-code.ts]: [P2] This retries every cleanup error and suppresses the final failure on Windows, not just the transient `EBUSY` described in the PR. Permanent failures can silently leak the per-call temp directory. Please use `fs.rm`'s built-in `maxRetries`/`retryDelay` support or retry only the explicitly supported transient error codes, and surface a persistent failure.

File: src/agent/loop.ts [hunk 4]
```diff
@@ -256,7 +256,7 @@ interface ParsedAction {
 }
 
 function parseAction(raw: string): ParsedAction | undefined {
-  const parsed = extractJsonObject<Record<string, unknown>>(raw);
+  const parsed = extractJsonObject<Record<string, unknown>>(repairUnbalancedObject(raw));
```
Comment (adshao) [inline src/agent/loop.ts]: [P2] Please split this parser behavior from the Windows transport fix and add focused tests before enabling it. Parsed actions can invoke framework tools, so accepting synthetically repaired model output is a trust-boundary change. Tests should prove that only missing tail closers are repaired, while mismatched delimiters, truncated strings, extra closers, and unrelated malformed JSON remain rejected; valid JSON must stay unchanged.

**Review by adshao — CHANGES_REQUESTED** [review]
Thanks for testing the provider on Windows and documenting the failures. Windows support is important, but this version cannot land because it changes the prompt trust boundary and introduces a shell-injection surface. It also bundles an independent model-output parser change into the platform fix without focused tests. Please address the inline blockers; preferably split the Windows transport work from JSON action repair so each can be reviewed and reverted independently.

**Comment by papisoV** [comment]
Thanks for the review — all four points addressed.

**P1, system-role boundary + command-line limit** — the rendered prompt is now written into the temp dir and passed with `--append-system-prompt-file`; stdin carries the user turn only. You were right that the stdin fold collapsed the boundary.

Verified against the installed CLI (2.1.223) on Windows: a probe whose answer depends on whether the instruction arrived in the system prompt or in the user message replies `SYSTROLE` — it is in the system role.

**P1, `shell: true` and the model argument** — the command line is now assembled here, with explicit cmd.exe quoting per argument, and the model id (the only caller-supplied value that reaches it) is validated against `^[A-Za-z0-9][A-Za-z0-9._:@/-]*$` before it can reach the shell. This also fixes two latent problems from letting Node concatenate the arguments: temp paths containing spaces survive intact, and the spawn no longer trips `DEP0190` ("arguments are not escaped, only concatenated").

I kept `claude.cmd` + shell because Windows installs the CLI as a `.cmd` shim and Node refuses to spawn one without a shell — but if you'd prefer resolving the underlying executable instead, I'm happy to switch.

**P2, rm retries** — replaced with `rm(tmp, { recursive: true, force: true, maxRetries: 5, retryDelay: 500 })`, which retries exactly the transient codes. A persistent failure is now reported to the run log as `claude_code_tmp_cleanup_failed` instead of being swallowed; it is not rethrown from `finally`, where it would mask the completion error already propagating out.

**P2, parser split** — the action repair is out of this PR and now lives in #105 with focused tests pinning both what is repaired and what stays rejected. The diff of this PR against `main` is `src/llm/claude-code.ts` only.

**Comment by adshao** [comment]
Thanks for the follow-up. The system/user prompt separation, cleanup retry behavior, and parser split are addressed. Three PR-specific items remain before this can land:

1. **[P1] Keep environment-derived paths out of the `cmd.exe` command string.** `quoteForCmd()` does not handle `%NAME%` expansion, while `systemPromptFile` is currently an absolute path derived from the OS temp directory. Since the child process already runs with `cwd: tmp`, please write the prompt to the absolute path but pass the constant relative argument `system-prompt.txt` to Claude. This removes the expansion edge case without requiring a more complex launcher.

2. **[P1] Add deterministic tests for the new Windows execution path.** Please cover unsafe model IDs, command construction, quoting and trailing backslashes, use of `--append-system-prompt-file`, preservation of the system/user boundary, and cleanup failure handling. A Windows CI job is preferred; alternatively, extract the command builder into a pure function that can be tested on every platform.

3. **[P2] Do not persist local absolute paths or raw filesystem errors.** The cleanup event currently records `dir: tmp` and the raw error message. Please omit or redact the path and log bounded fields such as the error code and syscall.

Please also update the PR description so it matches the current implementation.

Once these are addressed, ping me and I’ll re-review.

**Comment by papisoV** [comment]
All three addressed in ee15365, and the PR description now matches the implementation.

**[P1] Keep environment-derived paths out of the command string** — done as you suggested. The prompt is written to the absolute path under the temp dir and passed as the constant relative `system-prompt.txt`; the child already runs with `cwd: tmp`, so nothing OS-derived reaches the command line.

I verified this live rather than by construction, and it turned up a correction to my earlier comment. The "answer SYSTROLE if you are reading this as a system instruction" probe I used to claim the role boundary was restored does not discriminate: run repeatedly it answers `USRROLE` about two thirds of the time regardless of how the file is passed — relative or absolute. The evidence for the boundary is instead a constraint-override test: a constraint placed in the file at system level is honored 3/3, whereas the same text folded into the user turn is treated as an injected instruction and only sometimes obeyed. Both path forms deliver the file's content (the model recalls a nonce that exists only in that file), so the relative name changes nothing about the role boundary — it only removes the expansion surface.

**[P1] Deterministic tests** — `tests/claude-code-provider.test.mjs`, 9 tests. I took the pure-function route, so no Windows CI job is needed: `buildClaudeCommand(args, platform = process.platform)` takes the platform as a parameter, so the Windows path is exercised on every platform. Covered: command construction per platform; cmd.exe quoting with spaces, embedded quotes and trailing backslashes; unsafe model ids rejected both directly and through `buildClaudeInvocation` — which now runs before `mkdtemp`, so an unsafe id cannot leave a directory behind on its way out; `--append-system-prompt-file` carrying the constant relative name (no separator, no drive letter, no `%`); the system prompt never on the command line with only the user turn on stdin; and cleanup — silent on success, and on persistent failure reported rather than thrown with exactly `{ code, syscall }` logged, on both the logger and the stderr branches.

**[P2] No persisted paths or raw errors** — the cleanup event now carries only `code` and `syscall`; `dir` and the raw message are gone. Both absences are asserted in the tests.

Regression check: the full suite on Windows 11 is 546 tests / 519 pass / 27 fail with this change, versus 537 / 510 / 27 at the same commit without it. The 27 failures are byte-identical in both runs (Apple container, Docker, toolchain pinning, provider auth, Windows path handling); the +9 are the new file.

Ready for re-review.

**Review by adshao — APPROVED** [review]
